File: packages/@dicebear/core/src/utils/license.ts
import { Style } from '../types.js';
import * as _ from './escape.js';

export function xml(style: Style<any>): string {
  const sourceName = style.meta?.source?.name;
  const sourceUrl = style.meta?.source?.url;
  const creatorName = style.meta?.creator?.name;
  const licenseUrl = style.meta?.license?.url;
  const copyright = text(style);

  if (!sourceName && !creatorName && !sourceUrl && !licenseUrl && !copyright) {
    return '';
  }

  // https://nsteffel.github.io/dublin_core_generator/generator.html
  return (
    '<metadata' +
    ' xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"' +
    ' xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"' +
    ' xmlns:dc="http://purl.org/dc/elements/1.1/"' +
    ' xmlns:dcterms="http://purl.org/dc/terms/">' +
    '<rdf:RDF>' +
    '<rdf:Description>' +
    (sourceName ? `<dc:title>${_.xml(sourceName)}</dc:title>` : '') +
    (creatorName ? `<dc:creator>${_.xml(creatorName)}</dc:creator>` : '') +
    (sourceUrl ? `<dc:source xsi:type="dcterms:URI">${_.xml(sourceUrl)}</dc:source>` : '') +
    (licenseUrl ? `<dcterms:license xsi:type="dcterms:URI">${_.xml(licenseUrl)}</dcterms:license>` : '') +
    (copyright ? `<dc:rights>${_.xml(copyright)}</dc:rights>` : '') +
    '</rdf:Description>' +
    '</rdf:RDF>' +
    '</metadata>'
  );
}

export function text(style: Style<any>): string {
  let title = style.meta?.source?.name ? `„${style.meta?.source?.name}”` : 'Design';
  let creator = `„${style.meta?.creator?.name ?? 'Unknown'}”`;

  if (style.meta?.source) {
    title += ` (${style.meta.source})`;
  }

  let result = '';

  if (style.meta?.license?.name !== 'MIT' && style.meta?.creator?.name !== 'DiceBear' && style.meta?.source?.name) {
    result += 'Remix of ';
  }

  result += `${title} by ${creator}`;

  if (style.meta?.license?.name) {
    result += `, licensed under „${style.meta?.license?.name}”`;

    if (style.meta?.license?.url) {
      result += ` (${style.meta.license.url})`;
    }
  }

  return result;
}

File: src/go/core/internal/render/license.go
package render

import (
	"strings"

	"github.com/dicebear/dicebear-go/v11/internal/style"
)

// Builds attribution strings and embedded RDF/Dublin Core metadata from a
// style's meta block. Mirrors every other port,
// including the nullish creator fallback and the empty-string-as-absent
// treatment.

// nonEmptyPtr returns (value, true) only when p is non-nil and non-empty,
// mirroring the JS falsy checks (`!sourceName` is true for both a missing field
// and "").
func nonEmptyPtr(p *string) (string, bool) {
	if p != nil && *p != "" {
		return *p, true
	}
	return "", false
}

// licenseText returns a single-line attribution string, or "" when no
// attribution data is available.
func licenseText(m *style.Meta) string {
	if m == nil {
		return ""
	}

	sourceName, hasSource := nonEmptyPtr(m.Source.Name)
	sourceURL, hasSourceURL := nonEmptyPtr(m.Source.URL)
	creatorRaw := m.Creator.Name
	creatorName, hasCreator := nonEmptyPtr(creatorRaw)
	licenseName, hasLicense := nonEmptyPtr(m.License.Name)
	licenseURL, hasLicenseURL := nonEmptyPtr(m.License.URL)

	if !hasSource && !hasCreator && !hasLicense {
		return ""
	}

	title := "Design"
	if hasSource {
		title = "“" + sourceName + "”"
	}
	if hasSourceURL {
		title += " (" + sourceURL + ")"
	}

	// JS uses creatorName ?? 'Unknown' (nullish): an empty creator name stays
	// empty; only a missing one becomes "Unknown".
	creatorDisplay := "Unknown"
	if creatorRaw != nil {
		creatorDisplay = *creatorRaw
	}
	creator := "“" + creatorDisplay + "”"

	var b strings.Builder

	// Skip the "Remix of" prefix for MIT-licensed or DiceBear-original styles.
	licenseIsMIT := hasLicense && licenseName == "MIT"
	creatorIsDiceBear := hasCreator && creatorName == "DiceBear"
	if !licenseIsMIT && !creatorIsDiceBear && hasSource {
		b.WriteString("Remix of ")
	}

	b.WriteString(title + " by " + creator)

	if hasLicense {
		b.WriteString(", licensed under “" + licenseName + "”")
		if hasLicenseURL {
			b.WriteString(" (" + licenseURL + ")")
		}
	}

	return b.String()
}

// licenseXML builds an embedded <metadata> block with Dublin Core terms, or ""
// when no metadata fields are populated.
func licenseXML(m *style.Meta) string {
	rights := licenseText(m)

	if m == nil {
		return ""
	}

	title, hasTitle := nonEmptyPtr(m.Source.Name)
	creatorName, hasCreator := nonEmptyPtr(m.Creator.Name)
	sourceURL, hasSourceURL := nonEmptyPtr(m.Source.URL)
	licenseURL, hasLicenseURL := nonEmptyPtr(m.License.URL)

	if !hasTitle && !hasCreator && !hasSourceURL && !hasLicenseURL && rights == "" {
		return ""
	}

	var fields strings.Builder
	if hasTitle {
		fields.WriteString("<dc:title>" + escapeXML(title) + "</dc:title>")
	}
	if hasCreator {
		fields.WriteString("<dc:creator>" + escapeXML(creatorName) + "</dc:creator>")
	}
	if hasSourceURL {
		fields.WriteString(`<dc:source xsi:type="dcterms:URI">` + escapeXML(sourceURL) + "</dc:source>")
	}
	if hasLicenseURL {
		fields.WriteString(`<dcterms:license xsi:type="dcterms:URI">` + escapeXML(licenseURL) + "</dcterms:license>")
	}
	if rights != "" {
		fields.WriteString("<dc:rights>" + escapeXML(rights) + "</dc:rights>")
	}

	return `<metadata xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"` +
		` xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"` +
		` xmlns:dc="http://purl.org/dc/elements/1.1/"` +
		` xmlns:dcterms="http://purl.org/dc/terms/">` +
		`<rdf:RDF><rdf:Description>` + fields.String() + `</rdf:Description></rdf:RDF></metadata>`
}
